pkg/grpc/exec/exec.go

```diff
@@ -427,6 +470,20 @@ func (msg *MsgExitEventUnix) RetryInternal(ev notify.Event, timestamp uint64) (*
 	internal, parent := process.GetParentProcessInternal(msg.ProcessKey.Pid, timestamp)
 	var err error
 
+	if option.Config.EnableProcessAncestors && ev.GetAncestors() == nil &&
```

Comment:
I think we could, but i might have to reimplement it due to the process cleanup problem. I've described it in the comment above. Furthermore, I'm still not sure, what to do about that problem, as all my attempts to deal with it to this day were not very reliable. I would really appreciate an advice.

Reply:
ok, yea.. seems like a problem.. not sure yet how to deal with that

Reply:
I guess, i can make it so `getCleanupEvent` returns `MsgProcessCleanupEventUnix` only if `MsgExecveEventUnix` event's process has `clone` flag.
```go
func (msg *MsgExecveEventUnix) getCleanupEvent() *MsgProcessCleanupEventUnix {
	if msg.Unix.Msg.CleanupProcess.Ktime == 0 || strings.Contains(msg.Unix.Process.Flags, "clone") != true {
		return nil
	}
	return &MsgProcessCleanupEventUnix{
		PID:   msg.Unix.Msg.CleanupProcess.Pid,
		Ktime: msg.Unix.Msg.CleanupProcess.Ktime,
	}
}
```

Then in `MsgExecveEventUnix` event i can make it so `parent.RefInc("parent")` is called only, if `tetragonProcess` has any of `clone` / `procFS` flags.
```go
if parent != nil && (strings.Contains(tetragonProcess.Flags, "clone") == true ||
		strings.Contains(tetragonProcess.Flags, "procFS") == true) {
	parent.RefInc("parent")
}
```

So then in `MsgExitEventUnix` event we would have it so all `exec()` processes would have their refcnt set to 1, `clone()` - to 0 and an actual parent - to 2. So we'll just have to call `ancestor.RefDec("parent")` for all ancestors, returned by `GetAncestorProcessesInternal()` with condition `process.process.Pid.Value == tetragonProcess.Pid.Value`, where `process` is the process inside `GetAncestorProcessesInternal()` and `tetragonProcess` is the process from `MsgExitEventUnix` event. So basically, `GetAncestorProcessesInternal()` would look like this:
```go
// GetAncestorProcessesInternal returns a slice, representing a continuous sequence of ancestors of
// the process, including the immediate parent, that satisfy a given condition. The last element of
// the slice corresponds to the ancestor that no longer meets the given condition.
// The initial process, identified by the exedId parameter, must meet the condition as well.
func GetAncestorProcessesInternal(execId string, condition func(*ProcessInternal) bool) []*ProcessInternal {
	var process *ProcessInternal
	var err error

	if process, err = procCache.get(execId); err != nil {
		return nil
	}

	var ancestors []*ProcessInternal
	for condition(process) == true {
		if process, err = procCache.get(process.process.ParentExecId); err != nil {
			logger.GetLogger().WithError(err).WithField("id in event", execId).Debug("ancestor process not found in cache")
			return nil
		}
		ancestors = append(ancestors, process)
	}

	return ancestors
}
```

This approach has an obvious flaw, of course. If `exec()` ancestor with id=M is already removed from the process cache at that time for some reason, then we wouldn't be able to call `RefDec()` for any of the previous M ancestors.

Reply:
@tpapagian any idea on this?

Reply:
No, the approach i've suggested above won't solve all problems, either. If process (ID=1) creates a child process (ID=2) and immediately exits, and then it's child (ID=2) creates its own child (ID=3) and exits as well, then, by the end, the initial process (ID=1) will have it's refcnt set to 0 and, as a result, will be removed from the process cache. So collecting all ancestors for the last child (ID=3) and beyond won't be possible.

Not to mention that this approach looks more like an overcomplicated hack, than an actual good solution to the problem. I might have greatly underestimated the complexity of this task initially, yes. Have to come up with something better. Sorry for that.

Reply:
Checking that now.

Reply:
I haven't followed all the details of this PR yet but why do you need to care about `getCleanupEvent`?

I was thinking something like the following:
1. In exec events get all ancestors in `GetAncestorProcessesInternal` and check if this event should go inside the eventcache or not
  a. If not, call RefInc for the process and parent (as we already do now) and for all ancestors
  b. If yes, do something similar in the eventcache handling mechanism
2. In exit events call RefDec in the process and parent (as we already do now) and to all ancestors
3. All the other events happen between exec and exit so the refcnt of the related processes is not equal to zero so they will stay in the process cache until the exit happens.

Possibly we need to keep more metadata on which processes we have called refInc (possibly inside processInternal structure?) and several other details that I possibly miss now.

But the purpose of `getCleanupEvent` is to avoid having leftovers when a process execs. It should not matter if the refCnt is 1 or 2 or 3 as long as this > 0 (in order not to have the process evicted) and to make sure that refCnt goes to zero on exit to avoid keeping unnecessary processes in the process cache.

Do I miss anything on why this approach will not work? I will have a closer look on the code in that PR and in all comments possibly tomorrow.

Reply:
What i've described in the [latest PR comment](https://github.com/cilium/tetragon/pull/2938#issuecomment-2404680643) is related specifically to the situation, where we have successive execs each of which replaces the previous process instead of creating a child process with clone. So, we can have like 5 successive execs and only one exit. And, if i'll call RefInc in each exec for all ancestors, each ancestor will have it's refcnt incremented 5 times, including previous execs, that will not get an exit. And then exit will decrement them all only 1 time. So, on their exit those ancestors will not have their refcnt set to 0, because they were incremented more times than decremented.

I do agree on the idea of incrementing/decrementing refcnt of all ancestors, yes. I just need to make sure, that i do the same amount of increments as i do decrements. And so far all my attempts end up being overcomplicated, specifically because of successive exec chains.

Reply:
I will try to write down an example to understand exactly the situation and post that here. This may help to understand what we miss exactly.

Reply:
Let me provide my thoughts. Let's assume the following in order to make things work:
1. `getAncestors` from exec calls `RefInc` for each ancestor
2. `getAncestors` from exit calls `RefDec` for each ancestor
3. `getAncestors` from clone calls `RefDec` for each ancestor
4. `getCleanupEvent` calls `RefDec` for each ancestor (except from the process and parent that it currently does)

Initially we have 1 process:
```
init    /usr/init pid=1 exec_id=1 refCnt=1 
```
After calling `clone()` on process with pid=1:
```
        /usr/init pid=1 exec_id=1 refCnt=2 [+1 from clone]
clone   /usr/init pid=2 exec_id=2 refCnt=1 [+1 from clone]
```
After calling `clone()` on process with pid=2:
```
        /usr/init pid=1 exec_id=1 refCnt=3 [+1 from getAncestors]
        /usr/init pid=2 exec_id=2 refCnt=2 [+1 from clone]
clone   /usr/init pid=3 exec_id=3 refCnt=1 [+1 from clone]
```
After calling `exec("/usr/a")` on process with pid=3:
```
        /usr/init pid=1 exec_id=1 refCnt=3 [+1 from getAncestors | -1 from getCleanupEvent]
        /usr/init pid=2 exec_id=2 refCnt=2 [+1 from getAncestors | -1 from getCleanupEvent]
        /usr/init pid=3 exec_id=3 refCnt=1 [+1 from exec         | -1 from getCleanupEvent]
exec    /usr/a    pid=3 exec_id=4 refCnt=1 [+1 from exec]
```
After calling `clone()` on process with pid=3:
```
        /usr/init pid=1 exec_id=1 refCnt=4 [+1 from getAncestors]
        /usr/init pid=2 exec_id=2 refCnt=3 [+1 from getAncestors]
        /usr/init pid=3 exec_id=3 refCnt=2 [+1 from getAncestors]
        /usr/a    pid=3 exec_id=4 refCnt=2 [+1 from clone]
clone   /usr/a    pid=4 exec_id=5 refCnt=1 [+1 from clone]
```
After calling `exec("/usr/b")` on process with pid=4:
```
        /usr/init pid=1 exec_id=1 refCnt=4 [+1 from getAncestors | -1 from getCleanupEvent]
        /usr/init pid=2 exec_id=2 refCnt=3 [+1 from getAncestors | -1 from getCleanupEvent]
        /usr/init pid=3 exec_id=3 refCnt=2 [+1 from getAncestors | -1 from getCleanupEvent]
        /usr/a    pid=3 exec_id=4 refCnt=2 [+1 from getAncestors | -1 from getCleanupEvent]
        /usr/a    pid=4 exec_id=5 refCnt=1 [+1 from exec         | -1 from getCleanupEvent]
exec    /usr/b    pid=4 exec_id=6 refcnt=1 [+1 from exec] 
```
After calling `exit()` on process with pid=4:
```
        /usr/init pid=1 exec_id=1 refCnt=3 [-1 from getAncestors]
        /usr/init pid=2 exec_id=2 refCnt=2 [-1 from getAncestors]
        /usr/init pid=3 exec_id=3 refCnt=1 [-1 from getAncestors]
        /usr/a    pid=3 exec_id=4 refCnt=1 [-1 from getAncestors]
        /usr/a    pid=4 exec_id=5 refCnt=0 [-1 from exit] [DELETED]
exit    /usr/b    pid=4 exec_id=6 refcnt=0 [-1 from exit] [DELETED]
```
After calling `exit()` on process with pid=3:
```
        /usr/init pid=1 exec_id=1 refCnt=2 [-1 from getAncestors]
        /usr/init pid=2 exec_id=2 refCnt=1 [-1 from getAncestors]
        /usr/init pid=3 exec_id=3 refCnt=0 [-1 from exit] [DELETED]
exit    /usr/a    pid=3 exec_id=4 refCnt=0 [-1 from exit] [DELETED]
```
After calling `exit() `on process with pid=2:
```
        /usr/init pid=1 exec_id=1 refCnt=1 [-1 from exit]
exit    /usr/init pid=2 exec_id=2 refCnt=0 [-1 from exit] [DELETED]
```
After calling `exit()` on process with pid=1:
```
exit    /usr/init pid=1 exec_id=1 refCnt=0 [-1 from exit] [DELETED]
```

As I initially said, in order to make that work, we need to do some modifications on `getAncestors` and `getCleanupEvent`. I haven't also thought of how this will work in the case when we need to add something in the eventcache. But we will find out those details if needed. 

Does this make sense? Do I miss anything?

Reply:
It does make sense, yes. This was the first thing i've tried, and, unironically, i think it worked the best. I chose to try something else, because `getAncestors` and `getCleanupEvent` here do a lot of +1/-1 calls that basically just cancel each other. So we kind of doing most of the work for no actual value.

Plus, i couldn't call RefInc/RefDec for ancestors right away inside `GetAncestorProcessesInternal`, because at that moment i couldn't guarantee, that i will be able to collect all of them there. So i looped over all ancestors at least 2 times: first time to collect them all, seconds - to call RefInc/RefDec for each.

The more processes i had with lots of ancestors the less good it seemed. So i eventually decided to abandom that approach and try to at least call RefInc/RefDec only when i need to. Honestly, i'm not sure, if that was the right call.

I'll implement your suggestions and come back, thank you.

Reply:
> because getAncestors and getCleanupEvent here do a lot of +1/-1 calls that basically just cancel each other

Yes, I agree on that. But this is just a matter or increasing/decreasing a counter (i.e. minimal overhead) in order to make a simpler design. Making reference counting correct is a bin challenging so I would prioritize a simpler design here.  

> Honestly, i'm not sure, if that was the right call.

Not sure either to be honest. I just provided my thoughts on how this could work but I haven't spend time to understand all the details. So what I provided in my previous message could have issues as well when dealing with the eventcache etc. 

> I'll implement your suggestions and come back, thank you.

I would propose to spend some time first to understand all the details and please ask any questions here in case something does not make sense. Once everything is clear, you can proceed with the implementation.